src/lib/comms/sol-socket-impl-linux.c

```diff
@@ -437,85 +369,105 @@ sol_socket_linux_bind(struct sol_socket *socket, const struct sol_network_link_a
 }
 
 static int
-sol_socket_option_to_linux(enum sol_socket_option option)
+sol_socket_linux_set_read_monitor(struct sol_socket *s, bool on)
 {
-    switch (option) {
-    case SOL_SOCKET_OPTION_REUSEADDR:
-        return SO_REUSEADDR;
-    case SOL_SOCKET_OPTION_REUSEPORT:
-        return SO_REUSEPORT;
-    default:
-        SOL_WRN("Invalid option %d", option);
-        break;
-    }
+    bool ret;
+    struct sol_socket_linux *sock = (struct sol_socket_linux *)s;
 
-    return -1;
-}
+    SOL_NULL_CHECK(sock->on_can_read, -EINVAL);
 
-static int
-sol_socket_level_to_linux(enum sol_socket_level level)
-{
-    switch (level) {
-    case SOL_SOCKET_LEVEL_SOCKET:
-        return SOL_SOCKET;
-    case SOL_SOCKET_LEVEL_IP:
-        return IPPROTO_IP;
-    case SOL_SOCKET_LEVEL_IPV6:
-        return IPPROTO_IPV6;
-    default:
-        SOL_WRN("Invalid level %d", level);
-        break;
+    if (!sock->watch) {
+        sock->watch = sol_fd_add(sock->fd, SOL_FD_FLAGS_IN, on_socket_event, sock);
+        SOL_NULL_CHECK(sock->watch, -EBADF);
+        return 0;
     }
 
-    return -1;
-}
-
-static int
-sol_socket_linux_setsockopt(struct sol_socket *socket, enum sol_socket_level level,
-    enum sol_socket_option optname, const void *optval, size_t optlen)
-{
-    int l, option;
-    struct sol_socket_linux *s = (struct sol_socket_linux *)socket;
+    if (on)
+        ret = sol_fd_set_flags(sock->watch,
+            sol_fd_get_flags(sock->watch) | SOL_FD_FLAGS_IN);
+    else
+        ret = sol_fd_unset_flags(sock->watch, SOL_FD_FLAGS_IN);
 
-    l = sol_socket_level_to_linux(level);
-    option = sol_socket_option_to_linux(optname);
-
-    return setsockopt(s->fd, l, option, optval, optlen);
+    return ret ? 0 : -EBADF;
 }
 
 static int
-sol_socket_linux_getsockopt(struct sol_socket *socket, enum sol_socket_level level,
-    enum sol_socket_option optname, void *optval, size_t *optlen)
+sol_socket_linux_set_write_monitor(struct sol_socket *s, bool on)
 {
-    int ret, l, option;
-    socklen_t len = 0;
-    struct sol_socket_linux *s = (struct sol_socket_linux *)socket;
+    bool ret;
+    struct sol_socket_linux *sock = (struct sol_socket_linux *)s;
 
-    option = sol_socket_option_to_linux(optname);
-    l = sol_socket_level_to_linux(level);
+    SOL_NULL_CHECK(sock->on_can_write, -EINVAL);
 
-    ret = getsockopt(s->fd, l, option, optval, &len);
-    SOL_INT_CHECK(ret, < 0, ret);
+    if (!sock->watch) {
+        sock->watch = sol_fd_add(sock->fd, SOL_FD_FLAGS_OUT, on_socket_event, sock);
+        SOL_NULL_CHECK(sock->watch, -EBADF);
+        return 0;
+    }
 
-    *optlen = len;
-    return 0;
+    if (on)
+        ret = sol_fd_set_flags(sock->watch,
+            sol_fd_get_flags(sock->watch) | SOL_FD_FLAGS_OUT);
```

Comment:
this is strange, because set_flags and unset_flag are similar names and different behaviors. Would you do another PR that changes `unset_flags` to `add/remove` flags?